cdap-ui/app/directives/widget-container/widget-stream-properties/widget-stream-properties.js

```diff
@@ -0,0 +1,156 @@
+angular.module(PKG.name + '.commons')
+  .directive('myStreamProperties', function() {
+    return {
+      restrict: 'E',
+      scope: {
+        model: '=ngModel',
+        config: '=',
+        plugins: '='
+      },
+      templateUrl: 'widget-container/widget-stream-properties/widget-stream-properties.html',
+      controller: function($scope, myHelpers) {
+        $scope.options = $scope.config['schema-types'];
+        var defaultType = $scope.config['schema-default-type'] || $scope.options[0];
+
+        var watcher;
+        function removeWatcher() {
+          if (watcher) {
+            // deregister watch
+            watcher();
+            watcher = null;
+          }
+        }
+
+        $scope.$watch('plugins.format', function() {
+          // watch for changes
+          removeWatcher();
+
+          // do things based on format
+          if (['clf', 'syslog', ''].indexOf($scope.plugins.format) > -1) {
+            $scope.model = null;
+            $scope.fields = 'NOTHING';
+          } else if ($scope.plugins.format === 'avro'){
+            $scope.fields = 'AVRO';
+            watcher = $scope.$watch('avro', formatAvro, true);
+
+            if ($scope.model) {
+              $scope.avro.schema = JSON.parse($scope.model);
+            }
+          } else if ($scope.plugins.format === 'grok') {
+            $scope.fields = 'GROK';
+            $scope.model = null;
+          }
+
+          else {
+            $scope.fields = 'SHOW';
+            watcher = $scope.$watch('properties', formatSchema, true);
+          }
+
+        });
+
+        // Format model
+        function initialize() {
+          var schema = {};
+          $scope.avro = {};
+
+          if ($scope.model) {
+            try {
+              schema = JSON.parse($scope.model);
+            } catch (e) {
+              $scope.error = 'Invalid JSON string';
+            }
+          }
+
+          schema = myHelpers.objectQuery(schema, 'fields');
```

Comment:
Do we need ```objectQuery``` here? It seems to be only one field inside the object. Could avoid a function call.